File: content/examples/other/leetcode/problems/longest-palindromic-substring/example-2/solution.py
def longest_palindrome(s):
    best_l = 0
    best_r = 0

    def expand(left, right):
        nonlocal best_l, best_r
        while left >= 0 and right < len(s) and s[left] == s[right]:
            left -= 1
            right += 1
        l = left + 1
        r = right - 1
        if r - l > best_r - best_l:
            best_l, best_r = l, r

    for i in range(len(s)):
        expand(i, i)
        expand(i, i + 1)
    return s[best_l:best_r + 1]


if __name__ == "__main__":
    print(longest_palindrome("babad"))

File: content/examples/other/leetcode/problems/longest-palindromic-substring/example-2/solution.js
function longestPalindrome(s) {
  let bestL = 0;
  let bestR = 0;

  function expand(left, right) {
    while (left >= 0 && right < s.length && s[left] === s[right]) {
      left -= 1;
      right += 1;
    }
    // now (left,right) is one step beyond palindrome
    const L = left + 1;
    const R = right - 1;
    if (R - L > bestR - bestL) {
      bestL = L;
      bestR = R;
    }
  }

  for (let i = 0; i < s.length; i += 1) {
    expand(i, i);
    expand(i, i + 1);
  }
  return s.slice(bestL, bestR + 1);
}

if (require.main === module) {
  console.log(longestPalindrome("babad"));
}

module.exports = { longestPalindrome };
